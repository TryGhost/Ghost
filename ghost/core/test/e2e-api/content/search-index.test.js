const assert = require('assert/strict');

const models = require('../../../core/server/models');
const context = require('../../utils/fixtures/context');
const {agentProvider, fixtureManager, matchers} = require('../../utils/e2e-framework');
const {anyContentVersion, anyEtag, anyISODateTimeWithTZ, anyString} = matchers;

describe('Search Index Content API', function () {
    let agent;

    before(async function () {
        agent = await agentProvider.getContentAPIAgent();
        await fixtureManager.init('posts', 'api_keys');
        await agent.authenticate();
    });

    describe('fetchPosts', function () {
        const searchIndexPostMatcher = {
            id: anyString,
            slug: anyString,
            title: anyString,
            excerpt: anyString,
            url: anyString,
            updated_at: anyISODateTimeWithTZ,
            visibility: anyString
        };

        it('should return a list of posts', async function () {
            const res = await agent.get('search-index/posts')
                .expectStatus(200)
                .matchHeaderSnapshot({
                    'content-version': anyContentVersion,
                    etag: anyEtag
                })
                .matchBodySnapshot({
                    posts: new Array(11)
                        .fill(searchIndexPostMatcher)
                });

            // Explicitly double-check that expensive fields are not included
            const post = res.body.posts[0];
            assert.equal(post.html, undefined, 'html field should be not included in the response');
            assert.equal(post.plaintext, undefined, 'plaintext field should be not included in the response');
            assert.equal(post.mobiledoc, undefined, 'mobiledoc field should be not included in the response');
            assert.equal(post.lexical, undefined, 'lexical field should be not included in the response');
        });
    });

    describe('fetchAuthors', function () {
        const searchIndexAuthorMatcher = {
            id: anyString,
            slug: anyString,
            name: anyString,
            url: anyString,
            profile_image: anyString
        };

        it('should return a list of authors', async function () {
            await agent.get('search-index/authors')
                .expectStatus(200)
                .matchHeaderSnapshot({
                    'content-version': anyContentVersion,
                    etag: anyEtag
                })
                .matchBodySnapshot({
                    authors: new Array(2)
                        .fill(searchIndexAuthorMatcher)
                });
        });
    });

    describe('fetchTags', function () {
        const searchIndexTagMatcher = {
            id: anyString,
            slug: anyString,
            name: anyString,
            url: anyString
        };

        it('should return a list of tags', async function () {
            await agent.get('search-index/tags')
                .expectStatus(200)
                .matchHeaderSnapshot({
                    'content-version': anyContentVersion,
                    etag: anyEtag
                })
                .matchBodySnapshot({
                    tags: new Array(6)
                        .fill(searchIndexTagMatcher)
                });
        });

        it('does not return internal tags', async function () {
            const internalTag = await models.Tag.add({
                name: 'Internal Tag',
                slug: 'internal-tag',
                visibility: 'internal'
            }, context.internal);

            const tags = await agent.get('search-index/tags').expectStatus(200);

            assert.equal(tags.body.tags.find(tag => tag.id === internalTag.id), undefined, 'should not include internal tag');
        });
    });
});
