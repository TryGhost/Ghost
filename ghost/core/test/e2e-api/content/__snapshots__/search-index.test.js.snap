// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Index Content API fetchAuthors should return a list of authors 1: [body] 1`] = `
Object {
  "authors": Array [
    Object {
      "id": Any<String>,
      "name": Any<String>,
      "profile_image": Any<String>,
      "slug": Any<String>,
      "url": Any<String>,
    },
    Object {
      "id": Any<String>,
      "name": Any<String>,
      "profile_image": Any<String>,
      "slug": Any<String>,
      "url": Any<String>,
    },
  ],
}
`;

exports[`Search Index Content API fetchAuthors should return a list of authors 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "public, max-age=0",
  "content-length": "347",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Search Index Content API fetchPosts should return a list of posts 1: [body] 1`] = `
Object {
  "posts": Array [
    Object {
      "excerpt": Any<String>,
      "id": Any<String>,
      "slug": Any<String>,
      "title": Any<String>,
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000\\\\\\+\\\\d\\{2\\}:\\\\d\\{2\\}/,
      "url": Any<String>,
      "visibility": Any<String>,
    },
    Object {
      "excerpt": Any<String>,
      "id": Any<String>,
      "slug": Any<String>,
      "title": Any<String>,
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000\\\\\\+\\\\d\\{2\\}:\\\\d\\{2\\}/,
      "url": Any<String>,
      "visibility": Any<String>,
    },
    Object {
      "excerpt": Any<String>,
      "id": Any<String>,
      "slug": Any<String>,
      "title": Any<String>,
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000\\\\\\+\\\\d\\{2\\}:\\\\d\\{2\\}/,
      "url": Any<String>,
      "visibility": Any<String>,
    },
    Object {
      "excerpt": Any<String>,
      "id": Any<String>,
      "slug": Any<String>,
      "title": Any<String>,
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000\\\\\\+\\\\d\\{2\\}:\\\\d\\{2\\}/,
      "url": Any<String>,
      "visibility": Any<String>,
    },
    Object {
      "excerpt": Any<String>,
      "id": Any<String>,
      "slug": Any<String>,
      "title": Any<String>,
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000\\\\\\+\\\\d\\{2\\}:\\\\d\\{2\\}/,
      "url": Any<String>,
      "visibility": Any<String>,
    },
    Object {
      "excerpt": Any<String>,
      "id": Any<String>,
      "slug": Any<String>,
      "title": Any<String>,
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000\\\\\\+\\\\d\\{2\\}:\\\\d\\{2\\}/,
      "url": Any<String>,
      "visibility": Any<String>,
    },
    Object {
      "excerpt": Any<String>,
      "id": Any<String>,
      "slug": Any<String>,
      "title": Any<String>,
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000\\\\\\+\\\\d\\{2\\}:\\\\d\\{2\\}/,
      "url": Any<String>,
      "visibility": Any<String>,
    },
    Object {
      "excerpt": Any<String>,
      "id": Any<String>,
      "slug": Any<String>,
      "title": Any<String>,
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000\\\\\\+\\\\d\\{2\\}:\\\\d\\{2\\}/,
      "url": Any<String>,
      "visibility": Any<String>,
    },
    Object {
      "excerpt": Any<String>,
      "id": Any<String>,
      "slug": Any<String>,
      "title": Any<String>,
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000\\\\\\+\\\\d\\{2\\}:\\\\d\\{2\\}/,
      "url": Any<String>,
      "visibility": Any<String>,
    },
    Object {
      "excerpt": Any<String>,
      "id": Any<String>,
      "slug": Any<String>,
      "title": Any<String>,
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000\\\\\\+\\\\d\\{2\\}:\\\\d\\{2\\}/,
      "url": Any<String>,
      "visibility": Any<String>,
    },
    Object {
      "excerpt": Any<String>,
      "id": Any<String>,
      "slug": Any<String>,
      "title": Any<String>,
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000\\\\\\+\\\\d\\{2\\}:\\\\d\\{2\\}/,
      "url": Any<String>,
      "visibility": Any<String>,
    },
  ],
}
`;

exports[`Search Index Content API fetchPosts should return a list of posts 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "public, max-age=0",
  "content-length": "4387",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Search Index Content API fetchTags does not return internal tags 1: [body] 1`] = `
Object {
  "tags": Array [
    Object {
      "id": Any<String>,
      "name": Any<String>,
      "slug": Any<String>,
      "url": Any<String>,
    },
    Object {
      "id": Any<String>,
      "name": Any<String>,
      "slug": Any<String>,
      "url": Any<String>,
    },
    Object {
      "id": Any<String>,
      "name": Any<String>,
      "slug": Any<String>,
      "url": Any<String>,
    },
    Object {
      "id": Any<String>,
      "name": Any<String>,
      "slug": Any<String>,
      "url": Any<String>,
    },
    Object {
      "id": Any<String>,
      "name": Any<String>,
      "slug": Any<String>,
      "url": Any<String>,
    },
    Object {
      "id": Any<String>,
      "name": Any<String>,
      "slug": Any<String>,
      "url": Any<String>,
    },
  ],
}
`;

exports[`Search Index Content API fetchTags does not return internal tags 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "public, max-age=0",
  "content-length": "693",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Search Index Content API fetchTags should return a list of tags 1: [body] 1`] = `
Object {
  "tags": Array [
    Object {
      "id": Any<String>,
      "name": Any<String>,
      "slug": Any<String>,
      "url": Any<String>,
    },
    Object {
      "id": Any<String>,
      "name": Any<String>,
      "slug": Any<String>,
      "url": Any<String>,
    },
    Object {
      "id": Any<String>,
      "name": Any<String>,
      "slug": Any<String>,
      "url": Any<String>,
    },
    Object {
      "id": Any<String>,
      "name": Any<String>,
      "slug": Any<String>,
      "url": Any<String>,
    },
    Object {
      "id": Any<String>,
      "name": Any<String>,
      "slug": Any<String>,
      "url": Any<String>,
    },
    Object {
      "id": Any<String>,
      "name": Any<String>,
      "slug": Any<String>,
      "url": Any<String>,
    },
  ],
}
`;

exports[`Search Index Content API fetchTags should return a list of tags 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "public, max-age=0",
  "content-length": "693",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Accept-Encoding",
  "x-powered-by": "Express",
}
`;
